playbooks/robusta_playbooks/event_enrichments.py

```diff
@@ -0,0 +1,128 @@
+
+    # creating the finding before the rate limiter, to use the service key for rate limiting
+    finding = Finding(
+        title=f"{event.obj.reason} {event.obj.type} for {k8s_obj.kind} {k8s_obj.namespace}/{k8s_obj.name}",
+        description=event.obj.message,
+        source=FindingSource.KUBERNETES_API_SERVER,
+        severity=FindingSeverity.INFO
+        if event.obj.type == "Normal"
+        else FindingSeverity.DEBUG,
+        finding_type=FindingType.ISSUE,
+        aggregation_key=f"Kubernetes {event.obj.type} Event",
+        subject=FindingSubject(
+            k8s_obj.name,
+            FindingSubjectType.from_kind(k8s_obj.kind),
+            k8s_obj.namespace,
+            KubeObjFindingSubject.get_node_name(k8s_obj),
+        ),
+    )
+    event.add_finding(finding)
+
+
+@action
+def event_resource_events(event: EventChangeEvent, params: EventEnricherParams):
+    """
+    Given a Kubernetes event, gather all other events on the same resource in the near past
+    """
+    if not event.get_event():
+        logging.error(
+            f"cannot run event_resource_events on alert with no events object: {event}"
+        )
+        return
+    obj = event.obj.involvedObject
+    events_table = get_resource_events_table(
+        f"*Related Events*",
+        obj.kind,
+        obj.name,
+        obj.namespace,
+        included_types=params.included_types,
+        max_events=params.max_events,
+    )
+    if events_table:
+        event.add_enrichment([events_table], {SlackAnnotations.ATTACHMENT: True})
+
+
+@action
+def pod_events_enricher(event: PodEvent, params: EventEnricherParams):
+    """
+    Given a Kubernetes pod, fetch related events in the near past
+    """
+    pod = event.get_pod()
+    if not pod:
+        logging.error(
+            f"cannot run pod_events_enricher on alert with no pod object: {event}"
+        )
+        return
+
+    events_table_block = get_resource_events_table(
+        "*Pod events:*",
+        pod.kind,
+        pod.metadata.name,
+        pod.metadata.namespace,
+        included_types=params.included_types,
+        max_events=params.max_events,
+    )
+    if events_table_block:
+        event.add_enrichment([events_table_block], {SlackAnnotations.ATTACHMENT: True})
+
+
+@action
+def deployment_events_enricher(
+    event: DeploymentEvent, params: ExtendedEventEnricherParams
+):
+    """
+    Given a deployment, fetch related events in the near past.
+
+    Can optionally fetch events for related pods instead.
+    """
+    dep = event.get_deployment()
+    if not dep:
+        logging.error(
+            f"cannot run deployment_events_enricher on alert with no deployment object: {event}"
+        )
+        return
+
+    if params.dependent_pod_mode:
+        events_table_block = get_resource_events_table(
+            "*Related pod events:*",
+            "Pod",
+            None,
+            dep.metadata.namespace,
+            name_substring=dep.metadata.name,
```

Comment:
Since we're using the deployment name, and a pod type, I think this will match all the deployment pods together.
I'm afraid it might be spammy (ignoring the max_events for a sec), with a deployment with many pods.

With the replicas mismatch, with many replicas, chances are we won't get the problematic pod events, right?

That might also be the cause for the duplications

Maybe we should just get the matching non-running pods, and get their events?

Reply:
I think there will be cases where all pod crashes, and cases where only some (like with lack of enough resources, or affinity limitations etc)

Maybe we need a little more logic to decide what to show

Reply:
Maybe get related non-running pods, and choose max 2 pods, to display the pod events

Reply:
fixed